src/blockchain/bitcoin_rpc.rs

```diff
@@ -0,0 +1,392 @@
+        Self(client)
+    }
+    async fn index_is_synced<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        index: u32,
+        database: &mut D,
+    ) -> Result<bool, Error> {
+        let networks = vec![Network::Bitcoin, Network::Testnet, Network::Regtest];
+        let path = DerivationPath::from(vec![ChildNumber::Normal { index }]);
+        let last_script = database
+            .get_script_pubkey_from_path(ScriptType::External, &path)?
+            .unwrap();
+        let synced = networks
+            .iter()
+            .map(|network| {
+                let address = Address::from_script(&last_script, *network).unwrap();
+                let response = self.0.as_mut().unwrap().get_address_info(&address);
+                if let Ok(result) = response {
+                    if let Some(is_watchonly) = result.is_watchonly {
+                        return is_watchonly;
+                    }
+                }
+                false
+            })
+            .any(|x| x);
+        Ok(synced)
+    }
+
+    async fn needs_sync<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        database: &mut D,
+    ) -> Result<(bool, bool), Error> {
+        let synced = match database.get_last_index(ScriptType::External)? {
+            None => false,
+            Some(last_index) => self.index_is_synced(last_index, database).await?,
+        };
+        let needs_rescan = self.index_is_synced(0, database).await?;
+        Ok((!synced, needs_rescan))
+    }
+
+    async fn importmulti<D: BatchDatabase + DatabaseUtils>(
+        &mut self,
+        rescan_since: Option<u64>,
+        database: &mut D,
+    ) -> Result<(), Error> {
+        let (timestamp, rescan) = match rescan_since {
+            Some(timestamp) => (ImportMultiRescanSince::Timestamp(timestamp), Some(true)),
+            None => (ImportMultiRescanSince::Now, None),
+        };
+        self.0.as_mut().unwrap().import_multi(
+            &database
+                .iter_script_pubkeys(Some(ScriptType::External))?
+                .iter()
+                .map(|script_pubkey| ImportMultiRequest {
+                    script_pubkey: Some(ImportMultiRequestScriptPubkey::Script(&script_pubkey)),
+                    watchonly: Some(true),
+                    timestamp,
+                    ..Default::default()
+                })
+                .collect::<Vec<_>>(),
+            Some(&ImportMultiOptions { rescan }),
+        )?;
+        info!("Successfully imported addresses over bitcoin RPC");
+        Ok(())
+    }
+}
+
+impl Blockchain for BitcoinRpcBlockchain {
+    fn offline() -> Self {
+        BitcoinRpcBlockchain(None)
+    }
+
+    fn is_online(&self) -> bool {
+        self.0.is_some()
+    }
+}
+
+#[async_trait(?Send)]
+impl OnlineBlockchain for BitcoinRpcBlockchain {
+    async fn get_capabilities(&self) -> HashSet<Capability> {
+        vec![].into_iter().collect()
+    }
+
+    async fn setup<D: BatchDatabase + DatabaseUtils, P: Progress>(
```

Comment:
It would also be nice if this could create watch-only wallets in Bitcoin Core. Otherwise the caller needs it's own "setup" logic.

To do this Magical would need to know the watch-only wallet name:

```
pub struct BitcoinRpcBlockchain(Option<Client>, Option<String>);
```

Where the string is the watch-only wallet name? `None` would just use default wallet. Or we could not use an `Option` to force people to use a watch-only wallet, which seems like a better practice anyway. Then the `BitcoinRpcBlockchain.setup` could:

- Call `loadwallet`
  - If it succeeds, check if the node recognizes our addresses.
    - If yes, we're done.
    - If no, export addresses.
  - If it fails, call `createwallet` to create a new watch-only wallet.
    - If this succeeds, export addresses

Reply:
I implemented this, but didn't use `loadwallet` because [this RPC call isn't supported yet](https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/121)